src/main/java/ch/njol/skript/sections/SecFor.java

```diff
@@ -0,0 +1,154 @@
+package ch.njol.skript.sections;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.SkriptAPIException;
+import ch.njol.skript.classes.Changer;
+import ch.njol.skript.config.SectionNode;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.TriggerItem;
+import ch.njol.skript.lang.Variable;
+import ch.njol.skript.lang.util.ContainerExpression;
+import ch.njol.skript.registrations.Feature;
+import ch.njol.skript.util.Container;
+import ch.njol.skript.util.Container.ContainerType;
+import ch.njol.skript.util.LiteralUtils;
+import ch.njol.util.Kleenean;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.Map;
+
+@Name("For Each Loop (Experimental)")
+@Description("""
+	A specialised loop section run for each element in a list.
+	Unlike the basic loop, this is designed for extracting the key & value from pairs.
+	The loop element's key/index and value can be stored in a variable for convenience.
+	
+	When looping a simple (non-indexed) set of values, e.g. all players, the index will be the loop counter number."""
+)
+@Examples({
+	"for each {_player} in players:",
+	"\tsend \"Hello %{_player}%!\" to {_player}",
+	"",
+	"loop {_item} in {list of items::*}:",
+	"\tbroadcast {_item}'s name",
+	"",
+	"for each key {_index} in {list of items::*}:",
+	"\tbroadcast {_index}",
+	"",
+	"loop key {_index} and value {_value} in {list of items::*}:",
+	"\tbroadcast \"%{_index}% = %{_value}%\"",
+	"",
+	"for each {_index} = {_value} in {my list::*}:",
+	"\tbroadcast \"%{_index}% = %{_value}%\"",
+})
+@Since("INSERT VERSION")
+public class SecFor extends SecLoop {
+
+	static {
+		Skript.registerSection(SecFor.class,
+			"(for [each]|loop) [value] %~object% in %objects%",
+			"(for [each]|loop) (key|index) %~object% in %objects%",
+			"(for [each]|loop) [key|index] %~object% (=|and) [value] %~object% in %objects%"
```

Comment:
I'm not sure about this `=` syntax. Is there a reason you chose it?

Reply:
I think it was either from one of the issues/discussions suggesting this or I was copying from some other language, It's been about six months and I can't really remember.

Reply:
Oh I remember, I think there was a worry about `{x} and {y}` being confusingly-ambiguous (not in terms of parsing but for a user to read, maybe thinking it was looping two items at a time) between a regular list of expressions and this specific syntax where the `and` is literally part of it, and we wanted a less-ambiguous alternative that emphasised the keys were mapped to the values.
I'd rather it wasn't a symbol (=) although I don't mind it, but I couldn't think of a better single word for it.

Reply:
I think we should avoid symbols if possible. I don't think `for each {_a} = {_b} in {_c::*}` is great.
In fact, we might want to require the usage of the keywords `key|index` and `value` for this option. `for each {_a} and {_b} in {_c::*}` makes it seem like `c` is holding some sort of tuple.

Reply:
I'm fine with getting rid of =, how would you feel about `,` as an alternative? Python has e.g. `for key, value in dict.items()` and that's not terrible IMO.
I'm just worried that writing out the whole `for key {blah} and value {blah} in ...` is going to be more verbose than is necessary, when really we just need tome way to indicate one is the index representing the other.

Reply:
That would certainly be a better shorthand than `=`